### ISSUE TITLE
graceful plugin shutdown does not happen with ^c

### ISSUE THREAD

**Issue by Courey** [architect/architect#1479]
**Describe the issue**
- If you are using Remix
- and you create a plugin with a start and end 
- log in the end function (additionally log asynchronously)
- when you use ^c to terminate the process on mac, you would expect to see graceful termination of the plugins, but I do not see any logging from the plugin end()

**Steps to reproduce**
1. Go to [this repo and download the code](https://github.com/Courey/subprocess_signal) with an example plugin
2. Run `npm install`
3. Run `npm run dev`
4. terminate the process using ^c (on mac)
5. See that the plugin does not gracefully shut down. No logs are seen from the async end() function from the example plugin.

- you see the app build
- you see the sandbox start
- you see the sandbox running Sandbox startup plugin
- you press ^c to terminate the process (this uses SIGINT)
- you see the process abruptly end and you do NOT see any of the logs from the async end()

```
$ npm run dev

> dev
> remix dev -c "arc sandbox -e testing"

 💿  remix dev

 info  building...
 info  built (213ms)
sub-process pid: 36929
         App ⌁ remix-architect-app
      Region ⌁ us-west-2
     Profile ⌁ @aws profile / AWS_PROFILE not configured
     Version ⌁ Architect 10.16.3
         cwd ⌁ /Users/courey/dev/subprocess_signal

✓ Sandbox @http (HTTP API mode / Lambda proxy v2.0 format / live reload) routes
    any /* ................................ server

    http://localhost:3333

✓ Sandbox Started in 7ms
❤︎ Local environment ready!

⚬ Sandbox Running 1 Sandbox startup plugin
starting example sandbox plugin
✓ Sandbox Ran Sandbox startup plugin in 1ms
✓ Sandbox File watcher now looking for project changes
^C
```

**Expected behavior**
- you see the app build
- you see the sandbox start
- you see the sandbox running Sandbox startup plugin
- you press ^c to terminate the process (this uses SIGINT)
- you see the logs from the async end() in the order shown below
- you see "Sandbox Ran Sandbox shutdown plugin"
- you see the process end after graceful termination of all subprocesses and plugins

```
$ npm run dev

> dev
> npm-run-all build --parallel "dev:*"


> build
> remix build

 info  building... (NODE_ENV=production)
 info  built (315ms)

> dev:remix
> remix watch


> dev:arc
> cross-env NODE_ENV=development arc sandbox

Watching Remix app in development mode...
💿 Building...
         App ⌁ remix-architect-app
      Region ⌁ us-west-2
     Profile ⌁ @aws profile / AWS_PROFILE not configured
     Version ⌁ Architect 10.13.1
         cwd ⌁ /private/tmp/my-remix-app

✓ Sandbox @http (HTTP API mode / Lambda proxy v2.0 format / live reload) routes
    any /* ................................ server

    http://localhost:3333

✓ Sandbox Started in 9ms
❤︎ Local environment ready!

⚬ Sandbox Running 1 Sandbox startup plugin
starting example sandbox plugin
✓ Sandbox Ran Sandbox startup plugin in 7ms
✓ Sandbox File watcher now looking for project changes
💿 Rebuilt in 441ms
[REMIX DEV] D7AD53A4 ready
^C
⚬ Sandbox Running 1 Sandbox shutdown plugin
stopping example sandbox plugin
This log is after the async logs.
ONE SECOND
TWO SECOND
THREE SECOND
FOUR SECOND
FIVE SECOND
✓ Sandbox Ran Sandbox shutdown plugin in 5ms
```

**Screenshots**
I don't know if this is helpful or not, but this was one of the tests I ran on it and the results were unexpected.

<img width="1723" alt="only_exit_when_directly_siginted" src="https://github.com/architect/architect/assets/6545874/14ded76f-c246-40d3-9400-f12e5573406a">

**Desktop**
Please complete the following information (if appropriate):
 - OS: mac 13.4 Ventura chip: Apple M2 Max
 - Browser: not applicable

**Comment by Courey** [comment]
Interestingly SIGTERM still retains the desired behavior. So whatever is blocking the SIGINT is NOT blocking the SIGTERM.
<img width="1502" alt="SIGTERM_works_as_expected" src="https://github.com/architect/architect/assets/6545874/bf19283d-be44-41ab-92a8-55bd420f4a7f">

**Comment by ryanblock** [comment]
@Courey thank you so much for the sleuthing here, and especially for the repro repo. That really helps us jump right into the crux of the issue. 💕

So I did a bunch of digging, and this appears to be on the Remix / dependencies side of things. I'm not sure how we can help here.

`remix dev` spawns Sandbox using [`execa`](https://www.npmjs.com/package/execa) here: https://github.com/remix-run/remix/blob/main/packages/remix-dev/devServer_unstable/index.ts#L99-L111; they do not specify a `killSignal` option, but we would expect that to be `SIGTERM` per `execa`'s defaults ([ref](https://www.npmjs.com/package/execa#killsignal)).

They have some custom subprocess termination here, but it appears (at least here, on my machine) that a ctrl-c input is actually invoking `exit-hook` here: https://github.com/remix-run/remix/blob/main/packages/remix-dev/devServer_unstable/index.ts#L277 Tracing that `exit-hook` call, I see ctrl-c terminate the spawned Sandbox process via `SIGINT` with `130`.

I see `execa` 5.0.1 resolved in my local tree when I pulled down your project; when `execa` gets `exit-hook`'s termination, it simply does a `spawn.kill()` of the child process – which is, in this case, Sandbox: https://github.com/sindresorhus/execa/blob/427c5c2011e365a274dcca1162334a51c7218bfc/lib/kill.js#L101

When it does all this on my machine, no exit signal handlers are invoked; the process just immediately terminates. It makes sense you'd see different results by targeting the individual PID of the spawned process outside of `remix dev` though.

Thoughts on next steps here?

**Comment by brianleroux** [comment]
@ryanblock try this for a reduced case repo?

```javascript
export default {
  sandbox: {
    start() {
      console.log("starting example sandbox plugin");
    },
    async end() {
      console.log("stopping example sandbox plugin");
      setTimeout(()=>{console.log("won't run: ONE SECOND")}, 1000)
      setTimeout(()=>{console.log("won't run: TWO SECOND")}, 2000)
      setTimeout(()=>{console.log("won't run: THREE SECOND")}, 3000)
      setTimeout(()=>{console.log("won't run: FOUR SECOND")}, 4000)
      setTimeout(()=>{console.log("won't run: FIVE SECOND")}, 5000)
    },
  },
};```

**Comment by ryanblock** [comment]
I did see that! That bit of the repro case wound up being a red herring; the Sandbox CLI terminates the process after the `end()` method resolves; `await` the `setTimeout` calls and they'll all properly log before the process terminates:

```js
export default {
  sandbox: {
    start() {
      console.log("starting example sandbox plugin");
    },
    async end() {
      console.log("stopping example sandbox plugin");
      await new Promise(res => setTimeout(()=>{console.log("ONE SECOND");res()}, 1000))
      await new Promise(res => setTimeout(()=>{console.log("TWO SECOND");res()}, 1000))
      await new Promise(res => setTimeout(()=>{console.log("THREE SECOND");res()}, 1000))
      await new Promise(res => setTimeout(()=>{console.log("FOUR SECOND");res()}, 1000))
      await new Promise(res => setTimeout(()=>{console.log("FIVE SECOND");res()}, 1000))
    },
  },
}
```

@Courey let me know how you'd like to proceed, right now the reduced case points to Remix / dependency issues and not Sandbox.

**Comment by Courey** [comment]
There is[ this related issue](https://github.com/remix-run/remix/issues/6871) open in Remix. It looks like they would love your input! 

>In the meantime, would love any arc experts to help me figure out how best to interop with Vite dev server since that's what we're moving to anyway.

**Comment by ryanblock** [comment]
Also worth noting: `process.on('exit', ...)`, `process.on('SIGINT', ...)`, `process.on('SIGTERM', ...)` are all assumed to be synchronous operations, whereas `sandbox.end()` is an asynchronous API (which is what I think folks would expect it to be). 

Sandbox's CLI interface currently does not rely on exit event handlers, but instead intercepts ctrl-c and gracefully quits asynchronously (see: https://github.com/architect/sandbox/blob/main/src/cli/_stdin.js#L36-L52). Of course, there's always room for improvement!

**Comment by Courey** [comment]
I did notice that pattern, which is why I added signal handling just so I could log what signal was getting to the process. It is quite possible to monitor that in Remix alone as well, but I wanted to check from within the process as well just to be thorough. The things I did to handle signals were not production ready kinds of things, they were gnarly wrestling with code kinds of things. So you definitely don't want them. 🤣
